fix: various TTL improvements and tests improvements

rangoo94 opened the pull request:
Fix problems with TTL that caused the tests to be flaky, and add fixes and coverage for other TTL-related problems:

* Handle correctly TTL=0 for all backends (no expiration)
* Fix passing TTL for etcd and CosmosDB backends (as a wrong format was used)
* Add general tests for both expiring and non-expiring backends for most backends
   * DynamoDB and InMemory backends do not have expiration tests, because InMemory doesn't implement it, while DynamoDB is hard to cover locally (neither localstack or amazon/dynamodb-local support that well)

rangoo94 commented:
If you prefer, I can split this PR to 2 more independent:
* 1st commit solves the issues with tests,
* the rest of commits there focus on TTL=0 issues

mennanov commented on leakybucket.go:
```diff
@@ -541,8 +568,9 @@ func NewLeakyBucketMemcached(cli *memcache.Client, key string, ttl time.Duration
-	state := LeakyBucketState{}
+	var stateWithTTL leakyBucketStateWithExpirationTime
 	done := make(chan struct{}, 1)
+	t.casId = 0
```
Please add a comment explaining why setting casId to zero is needed here.

Btw, is it really needed? If the read operation succeeds then `casId` will be set to `item.CasID` below, if the read operation fails then the last successfully read `casId` value will be kept (as it probably should). Am i missing something?

mennanov commented on leakybucket.go:
```diff
@@ -552,25 +580,29 @@ func (t *LeakyBucketMemcached) State(ctx context.Context) (LeakyBucketState, err
 	}
 	b := bytes.NewBuffer(item.Value)
-	err = gob.NewDecoder(b).Decode(&state)
+	err = gob.NewDecoder(b).Decode(&stateWithTTL)
```
I am concerned about this change: the previously stored states (which are `LeakyBucketState` struct) will likely fail to Decode which will break the existing limiters.

Is this actually needed? Does Memcached not enforce the TTL parameter strictly?

mennanov replied:
I'm not sure this additional check for Memcached is well justified: 

1. it slightly increases complexity
2. makes Memcached implementation inconsistent with other backends
3. realistically - should we care that much about a second precision? 

What do you think?

rangoo94 replied:
> makes Memcached implementation inconsistent with other backends

It's actually making it more compatible with other backends.

| Backend | Description | Example: item added at 00:00.333 | Example: item added at 00:00.999 |
|-|-|-|-|
| **CosmosDB** | Expiration is set in duration **Bug: right now timestamp is sent so it's deleted in 50 years** | - | - |
| **DynamoDB** | Expiration is set in timestamp (seconds) - eligible for async deletion | Expires anytime after 00:01.000 | Expires anytime after 00:01.000 |
| **Memcached** | Expiration is set in timestamp (seconds) - deleted when reached | Expires at 00:01.000 | Expires at 00:01.000 |
| **Redis** | Expiration is precisely set in duration | Expires at 00:01.333 | Expires at 00:01.999 |
| **Memcached (after changes)** | Expires at 00:01.333 | Expires at 00:00.999 |

So:
* This makes it same as Redis
* This makes it more similar to DynamoDB **(although it may be worth to add the same mechanism there)**
* CosmosDB implementation has a bug, that is basically keeping the entries indefinitely
   * It works fine though otherwise, as there is a  `time.Now().Unix() > item.TTL`, similarly to one in Memcached

Definitely it's worth to do that for Memcached, and it would be good to fix both CosmosDB and DynamoDB implementations.

> realistically - should we care that much about a second precision?

For the lower TTLs, like 1 second, it could make the item expire immediately, what is definitely not expected behavior - confirmed by existing tests. Even for higher TTLs, it could leak unexpectedly. As an example, someone could set:
* capacity=3, rate=1s, ttl=3s
* it should always keep it at 3 ops/3s, as the `TTL = capacity * rate`
* in practice, it may result in up to 3 ops/2s (50% higher load), as the key may expire early

mennanov replied:
I prefer option 2: making the TTL a responsibility of a developer and the chosen backend. This library should not try to "fix" the TTL mechanisms of the underlying backends because pragmatically the TTL parameter should be large enough to never interfere with the limiters behavior (e.g. `capacity * rate * some_large_enough_number`) and instead should be used to merely reclaim unused space.

I see the following action items:

- [ ] fix the  CosmosDB implementation to properly set the TTL parameter (seconds to live instead of a timestamp) and remove the manual check.
- [ ] update the tests in such a way that they allow enough time for the Cosmos DB TTL Garbage Collector to kick in.

rangoo94 replied:
> update the tests in such a way that they allow enough time for the Cosmos DB TTL Garbage Collector to kick in.

There are no TTL tests yet for expiration of other sources than Memcached. I fixed the tests/implementation for the Memcached and CosmosDB though.

I checked if it's possible to make it general for all backends, but:
* DynamoDB & etc - for some reason it's not expiring at all (but I haven't investigated enough to see if it's `dynamo-local`, setup, or implementation issue
* InMemory - has no implementation for TTL

Aside, seems like Bitnami has deprecated some of their images, so I replaced the non-hardened images with their alternative.

rangoo94 replied:
Thank you! Sorry for the delay, but I had busy time a bit.

I've checked the tests and:
* I updated TTLs in regular tests to higher values to not mix them there
* I fixed Etcd implementation - TTL was sent in nanoseconds, while it expects seconds
* I added `*NoExpiration` (ttl=0) tests for all buckets
* I added `*TTLExpiration` (ttl=1s, wait=3s) tests for most of buckets. I added exceptions for:
  * **in-memory backend:** it has no TTL built-in
  * **DynamoDB:** `amazon/dynamodb-local` doesn't clean data at all, `localstack` would clean it once per hour, so mimicking it may be problematic

It should be better now 🙂

mennanov commented:
@leeym since this PR touches DynamoDB i'd like you to take a look, thanks!

mennanov commented:
@rangoo94 great job! Overall looks great, i left a few comments to clear things up. Thanks!

mennanov commented on leakybucket.go:
```diff
@@ -238,11 +243,18 @@ func (l *LeakyBucketEtcd) createLease(ctx context.Context) error {
+		clientv3.OpPut(etcdKey(l.prefix, etcdKeyLBLast), fmt.Sprintf("%d", state.Last), opts...),
+	}
+	if l.ttl > 0 {
+		ops = append(ops, clientv3.OpPut(etcdKey(l.prefix, etcdKeyTBLease), fmt.Sprintf("%d", l.leaseID), opts...))
```
Should `etcdKeyTBLease` be `etcdKeyLBLease`?

mennanov commented on leakybucket_test.go:
```diff
@@ -143,11 +144,64 @@ func (s *LimitersTestSuite) TestLeakyBucketOverflow() {
+	// Ignore in-memory bucket, as it has no expiration,
+	// ignore DynamoDB, as amazon/dynamodb-local doesn't support TTLs.
+	for k := range buckets {
+		if strings.Contains(k, "BucketInMemory") || strings.Contains(k, "BucketDynamoDB") {
```
To avoid iterating over the buckets twice let's move this check to the for-loop below and skip InMemory and DynamoDB backend with `continue`

rangoo94 replied:
Actually, I'm iterating 3 times:
* Ignore in-memory & DynamoDB buckets
* Fill all the buckets
* Check if the buckets are empty

If I skip this, I would either need to:
* **do this check in both loops:** which may be less clear
* **fill buckets and clear them in a single loop:** this would increase the test duration, as it works on a real clock - now it's `3 seconds`, while then it would be `number_of_buckets * 3 seconds` - 11 backends, 7 lockers = 231 seconds = ~4 minutes for this test + the same for TokenBucket one

Personally, I think it's the best way, but if you prefer I can adjust it in either way 👍

mennanov commented on leakybucket_test.go:
```diff
@@ -143,11 +144,64 @@ func (s *LimitersTestSuite) TestLeakyBucketOverflow() {
+	// Wait for 3 seconds to check if the items have been deleted successfully
+	clock.Sleep(3 * time.Second)
+
+	// Expect all buckets to be still filled
```
Is this comment accurate? It should be the opposite, right?

mennanov commented on leakybucket_test.go:
```diff
@@ -143,11 +144,64 @@ func (s *LimitersTestSuite) TestLeakyBucketOverflow() {
 	capacity := int64(2)
 	clock := newFakeClock()
-	for name, bucket := range s.leakyBuckets(capacity, rate, clock) {
+	for name, bucket := range s.leakyBuckets(capacity, rate, time.Minute, clock) {
```
Since TTL is irrelevant  for this test i'd just set it to `0` to never expire.

mennanov commented on leakybucket_test.go:
```diff
@@ -48,7 +49,7 @@ func (s *LimitersTestSuite) TestLeakyBucketRealClock() {
 	clock := l.NewSystemClock()
 	for _, requestRate := range []time.Duration{rate / 2} {
-		for name, bucket := range s.leakyBuckets(capacity, rate, clock) {
+		for name, bucket := range s.leakyBuckets(capacity, rate, time.Minute, clock) {
```
Here and in other similar places below (and other test files): if a TTL is irrelevant for this test does it make sense to set it to `0`?

mennanov commented on tokenbucket_test.go:
```diff
@@ -316,14 +279,65 @@ func (s *LimitersTestSuite) TestTokenBucketRedisBackwardCompatibility() {
+	// Ignore in-memory bucket, as it has no expiration,
+	// ignore DynamoDB, as amazon/dynamodb-local doesn't support TTLs.
+	for k := range buckets {
+		if strings.Contains(k, "BucketInMemory") || strings.Contains(k, "BucketDynamoDB") {
```
Same: let's move it to the for-loop below

mennanov commented on tokenbucket_test.go:
```diff
@@ -316,14 +279,65 @@ func (s *LimitersTestSuite) TestTokenBucketRedisBackwardCompatibility() {
+	// Wait for 3 seconds to check if the items have been deleted successfully
+	clock.Sleep(3 * time.Second)
+
+	// Expect all buckets to be still filled
```
Same: please, update the comment

rangoo94 commented:
> Please, update the description of this PR to better reflect the code changes. Also, please, update the branch (merge main or rebase).

Thanks @mennanov, I updated the PR description, adjusted the invalid comments, and rebased on top of `master`. 👍
